src/netbase.cpp

```diff
@@ -40,49 +42,71 @@ bool fNameLookup = DEFAULT_NAME_LOOKUP;
+    }
+    res->complete = true;
+}
+
+std::vector<CNetAddr> DNSResolveAsync(const std::string& name, bool allow_lookup)
+{
+    raii_event_base base = obtain_event_base();
+    raii_evdns_base dnsbase = obtain_evdns_base(base.get());
+
+    struct evutil_addrinfo hints{};
     // We want a TCP port, which is a streaming socket type
-    ai_hint.ai_socktype = SOCK_STREAM;
-    ai_hint.ai_protocol = IPPROTO_TCP;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_protocol = IPPROTO_TCP;
     // We don't care which address family (IPv4 or IPv6) is returned
-    ai_hint.ai_family = AF_UNSPEC;
-    // If we allow lookups of hostnames, use the AI_ADDRCONFIG flag to only
+    hints.ai_family = AF_UNSPEC;
+    // If we allow lookups of hostnames, use the EVUTIL_AI_ADDRCONFIG flag to only
     // return addresses whose family we have an address configured for.
     //
-    // If we don't allow lookups, then use the AI_NUMERICHOST flag for
+    // If we don't allow lookups, then use the EVUTIL_AI_NUMERICHOST flag for
     // getaddrinfo to only decode numerical network addresses and suppress
     // hostname lookups.
-    ai_hint.ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+    hints.ai_flags = allow_lookup ? EVUTIL_AI_ADDRCONFIG : EVUTIL_AI_NUMERICHOST;
 
-    addrinfo* ai_res{nullptr};
-    const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
-    if (n_err != 0) {
-        return {};
+    std::vector<CNetAddr> resolved_addresses;
+    struct DNSResolveResponse res = {.complete = false, .vAddr = &resolved_addresses};
+
+    auto req = evdns_getaddrinfo(dnsbase.get(), name.c_str(), nullptr, &hints, GAICallback, &res);
+
+    if (!req) {
+        // evdns_getaddrinfo() returned immediately.
+        // There might have been an error, or no name lookup was necessary
+        return resolved_addresses;
     }
 
-    // Traverse the linked list starting with ai_trav.
-    addrinfo* ai_trav{ai_res};
-    std::vector<CNetAddr> resolved_addresses;
-    while (ai_trav != nullptr) {
-        if (ai_trav->ai_family == AF_INET) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
-        }
-        if (ai_trav->ai_family == AF_INET6) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in6));
-            const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
-            resolved_addresses.emplace_back(s6->sin6_addr, s6->sin6_scope_id);
-        }
-        ai_trav = ai_trav->ai_next;
+    // We need to wait for the name resolution callback on a new temporary event loop
+    std::thread thread(event_base_loop, base.get(), 1);
+
+    int checks = 0;
+    while (!res.complete)
+    {
+        // Check every 100ms for 2 seconds
+        if (++checks > 20) break;
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
     }
-    freeaddrinfo(ai_res);
 
+    thread.join();
```

Comment:
2 sec timeout waiting for the callback and then the loop will close. With my black hole dns resolver, the patch saves 30 seconds against master on the unit test in the second commit. 

It took me some work to figure out that using ..._once() to start the event loop would allow me to quit the loop manually which allows join() to succeed

Reply:
(libevent does not call the platform getaddrinfo for dns requests, it has its own async methods)

Reply:
> libevent does not call the platform getaddrinfo for dns requests, it has its own async methods

That explains why it doesn't hang, thanks!

But it does seem like it uses `getaddrinfo` in some cases: https://github.com/libevent/libevent/blob/75208132d5b7a8fff59ca3bf47253179ec314951/evutil.c#L1686

Reply:
Yeah I think thats just to decode numeric addresses like "100.200.30.40" etc. It looked to me like if an actual DNS request is necessary it has its own methods to make reqests directly to the nameservers provided by the platform? When things get "serious":

https://github.com/libevent/libevent/blob/master/evdns.c#L5623